advanced/dynamic-tcp-client/src/main/java/org/springframework/integration/samples/dynamictcp/DynamicTcpClientApplication.java

```diff
@@ -0,0 +1,162 @@
+	@Bean
+	public IntegrationFlow toTcp() {
+		return f -> f.route(router());
+	}
+
+	@Bean
+	public TcpRouter router() {
+		return new TcpRouter();
+	}
+
+	// Two servers
+
+	@Bean
+	public TcpReceivingChannelAdapter inOne(TcpNetServerConnectionFactory cfOne) {
+		TcpReceivingChannelAdapter adapter = new TcpReceivingChannelAdapter();
+		adapter.setConnectionFactory(cfOne);
+		adapter.setOutputChannel(outputChannel());
+		return adapter;
+	}
+
+	@Bean
+	public TcpNetServerConnectionFactory cfOne() {
+		return new TcpNetServerConnectionFactory(1234);
+	}
+
+	@Bean
+	public TcpReceivingChannelAdapter inTwo(TcpNetServerConnectionFactory cfTwo) {
+		TcpReceivingChannelAdapter adapter = new TcpReceivingChannelAdapter();
+		adapter.setConnectionFactory(cfTwo);
+		adapter.setOutputChannel(outputChannel());
+		return adapter;
+	}
+
+	@Bean
+	public TcpNetServerConnectionFactory cfTwo() {
+		return new TcpNetServerConnectionFactory(5678);
+	}
+
+	@Bean
+	public QueueChannel outputChannel() {
+		return new QueueChannel();
+	}
+
+	public static class TcpRouter extends AbstractMessageRouter {
+
+		private final static int MAX_CACHED = 10; // When this is exceeded, we remove the LRU.
+
+		@SuppressWarnings("serial")
+		private final LinkedHashMap<String, MessageChannel> subFlows =
+				new LinkedHashMap<String, MessageChannel>(MAX_CACHED, .75f, true) {
+
+					@Override
+					protected boolean removeEldestEntry(Entry<String, MessageChannel> eldest) {
+						if (size() > MAX_CACHED) {
+							removeSubFlow(eldest);
+							return true;
+						}
+						else {
+							return false;
+						}
+					}
+
+				};
+
+		@Autowired
+		private ConfigurableApplicationContext context;
+
+		@Autowired
+		private IntegrationFlowContext flowContext;
+
+		@Override
+		protected synchronized Collection<MessageChannel> determineTargetChannels(Message<?> message) {
+			MessageChannel channel = this.subFlows
+					.get("" + message.getHeaders().get("host") + message.getHeaders().get("port"));
+			if (channel == null) {
+				channel = createNewSubflow(message);
+			}
+			return Collections.singletonList(channel);
+		}
+
+		private MessageChannel createNewSubflow(Message<?> message) {
+			String host = (String) message.getHeaders().get("host");
+			Integer port = (Integer) message.getHeaders().get("port");
+			Assert.state(host != null && port != null, "host and/or port header missing");
+			String hostPort = host + port;
+
+			TcpNetClientConnectionFactory cf = new TcpNetClientConnectionFactory(host, port);
+			this.context.getBeanFactory().registerSingleton(hostPort + ".cf", cf);
+			this.context.getBeanFactory().initializeBean(cf, hostPort+".cf");
+			cf.afterPropertiesSet();
+			TcpSendingMessageHandler handler = new TcpSendingMessageHandler();
+			handler.setConnectionFactory(cf);
+			IntegrationFlow flow = f -> f.handle(handler);
+			this.flowContext.register(hostPort + ".flow", flow);
+			MessageChannel channel = context.getBean(hostPort + ".flow.input", MessageChannel.class);
```

Comment:
Yes, thanks, didn't notice that method and yes, I agree with the other discussion.